### PR TITLE
Implementation of the call api

### PR DESCRIPTION
**Pull request by ezequielsetzesglobant**
The call to the api with retrofit was implemented together with the dependency injections and the coroutines.

Application
https://user-images.githubusercontent.com/84338513/131723571-72addfe0-f1a4-452b-84b9-506b47b64f5d.mp4

Structure project
image
image
image
image
image

PostMan
image
Android Studio
image

### CODE REVIEW

File: app/src/main/java/com/example/cleanarchitecture/activity/MainActivity.kt [hunk 1]
```diff
@@ -1,12 +1,45 @@
...
+
+        viewModel.liveDataState.observe(this, { updateUI(it) })
+
+        binding.buttonMainActivityIdFetch.setOnClickListener { viewModel.onFetchCharactersClicked() }
+    }
+
+    private fun updateUI(characterData: Event<Data<CharacterDataWrapperEntity>>?) {
+        when (characterData?.peekContent()?.responseType) {
+            Status.SUCCESSFUL -> {
+                Toast.makeText(
```
Comment (BQuinteros) [inline app/src/main/java/com/example/cleanarchitecture/activity/MainActivity.kt:33]: here you can do a fun how show this toast passing the string in the parameter

File: app/src/main/java/com/example/cleanarchitecture/utils/Data.kt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package com.example.cleanarchitecture.utils
+
+data class Data<RequestData>(var responseType: Status, var data: RequestData? = null, var error: Exception? = null)
```
Comment (SartiMau) [inline app/src/main/java/com/example/cleanarchitecture/utils/Data.kt:3]: as a good practice, each viewModel should have its own data and status, please put these classes in the ViewModel that use them and remove this file

File: app/src/main/java/com/example/cleanarchitecture/activity/MainActivity.kt [hunk 3]
```diff
@@ -1,12 +1,45 @@
...
+                    this,
+                    getString(R.string.toast_main_activity_characters, characterData.peekContent().data?.data?.total),
+                    Toast.LENGTH_LONG
+                ).show()
+            }
+            Status.ERROR -> {
+                Toast.makeText(this, characterData.peekContent().error.toString(), Toast.LENGTH_LONG).show()
+            }
+            else -> return
```
Comment (SartiMau) [inline app/src/main/java/com/example/cleanarchitecture/activity/MainActivity.kt:42]: do not use else here, add the other status and give them a functionality

File: app/src/main/java/com/example/cleanarchitecture/viewmodel/CharacterViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+class CharacterViewModel(private val characterUseCase: CharacterUseCase) : ViewModel() {
+
+    private var mutableCharactersState: MutableLiveData<Event<Data<CharacterDataWrapperEntity>>> = MutableLiveData()
+    val liveDataState: LiveData<Event<Data<CharacterDataWrapperEntity>>>
+        get() = mutableCharactersState
+
+    fun onFetchCharactersClicked() {
+        viewModelScope.launch {
+            mutableCharactersState.value = Event(Data(responseType = Status.LOADING))
```
Comment (SartiMau) [inline app/src/main/java/com/example/cleanarchitecture/viewmodel/CharacterViewModel.kt:26]: the view is doing nothing for this state

File: domain/src/main/java/com/example/domain/usecase/CharacterUseCase.kt [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package com.example.domain.usecase
+
+import com.example.domain.entity.CharacterDataWrapperEntity
+import com.example.domain.utils.Result
+
+interface CharacterUseCase {
```
Comment (SartiMau) [inline domain/src/main/java/com/example/domain/usecase/CharacterUseCase.kt:6]: for the useCase you can put in the same file the interface and the implementation

File: domain/src/main/java/com/example/domain/usecase/CharacterUseCaseImpl.kt [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package com.example.domain.usecase
+
+import com.example.domain.entity.CharacterDataWrapperEntity
+import com.example.domain.repository.CharacterRepository
+import com.example.domain.utils.Result
+
+class CharacterUseCaseImpl(private val characterRepository: CharacterRepository) : CharacterUseCase {
```
Comment (SartiMau) [inline domain/src/main/java/com/example/domain/usecase/CharacterUseCaseImpl.kt:7]: for the useCase you can put in the same file the interface and the implementation

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/cleanarchitecture/activity/MainActivity.kt (hunk 1 above)
Comment (ezequielsetzesglobant) [reply app/src/main/java/com/example/cleanarchitecture/activity/MainActivity.kt:33]: done

File: app/src/main/java/com/example/cleanarchitecture/utils/Data.kt (hunk 2 above)
Comment (ezequielsetzesglobant) [reply app/src/main/java/com/example/cleanarchitecture/utils/Data.kt:3]: done

File: domain/src/main/java/com/example/domain/usecase/CharacterUseCase.kt (hunk 5 above)
Comment (ezequielsetzesglobant) [reply domain/src/main/java/com/example/domain/usecase/CharacterUseCase.kt:6]: done

File: domain/src/main/java/com/example/domain/usecase/CharacterUseCaseImpl.kt (hunk 6 above)
Comment (ezequielsetzesglobant) [reply domain/src/main/java/com/example/domain/usecase/CharacterUseCaseImpl.kt:7]: done

File: app/src/main/java/com/example/cleanarchitecture/viewmodel/CharacterViewModel.kt (hunk 4 above)
Comment (ezequielsetzesglobant) [reply app/src/main/java/com/example/cleanarchitecture/viewmodel/CharacterViewModel.kt:26]: done

**Review by SartiMau — APPROVED** [review]

**Review by lucianoeli — APPROVED** [review]

**Review by BQuinteros — APPROVED** [review]

**Pull request merged**